### ISSUE TITLE
Trying to ignore some fields when serializing/deserializing an object graph 

### ISSUE THREAD

**Issue by Bram80** [RuedigerMoeller/fast-serialization#75]
Hi,

I'm trying to clone (deep copy) a complete object graph fetched from DB using Hibernate.
The thing is that each entity should be considered as a new one ('transient' in Hibernate lingo).
Therefore I had the plan to set each entity's ID to 'null' and it's version property to zero.

My setup is currently like this:

first the global config:
![image](https://cloud.githubusercontent.com/assets/5504516/8531876/96489baa-242a-11e5-9458-927220d2f9bf.png)

Some 'hook' to ensure that we're working with a 'real' object instead of a Hibernate proxy:
![image](https://cloud.githubusercontent.com/assets/5504516/8531883/ac00ae9c-242a-11e5-907c-25626c932e8e.png)

Set a callback to erase each ID and version property for each entity that contains a @Conditional annotation:
![image](https://cloud.githubusercontent.com/assets/5504516/8531893/c10e4c68-242a-11e5-97a7-a53bf6245656.png)

The 'ProxySafeObjectOutput' is working fine; However, I'm experiencing issues with the callback that is invoked upon deserializing the object graph. Almost each entity in the graph contains a reference to an other one; it also contains lot of cyclic references where each reference is annotated with '@Conditional'.
I'm constantly having 'NullpointerException's upon deserialing; I'm sure that it's related to the '@Conditional' annotation because if I remove it everything works (besides resetting Id and version fields).

Is there an other way (some hook perhaps) that I can use to check if an entity field should be skipped?

I'm invoking the deep clone as follows:
![image](https://cloud.githubusercontent.com/assets/5504516/8532152/893023a0-242c-11e5-99f9-1c80da178bc2.png)

Any ideas?

**Comment by RuedigerMoeller** [comment]
Hm .. advanced chemistry :) . I have to take a deeper look once back in my nerd cavern (at office currently).
on a first glance:
- your check for proxies will be bypassed for classes where a serializer is registered (probably is not the case but still a caveat)
- conditional does not have deep testcases and was implemented long ago (1.x), so it might be broken for complex objectgraphs.
- another risk (unsure have to check the code) is that reference sharing might be distorted as the lookup will be done using the proxy object, registration migth happen using the unproxied object.

Before investigating workaround's I'd verify conditional in general. However in either case this problem will be solvable.

Thanks for detailed report :)

**Comment by RuedigerMoeller** [comment]
First the bad news (I could reproduce similar issues in a test):

To avoid writing fq classnames for the same class several times, fst enumerates classes encoutered during write/read. Once a classname is written second time, the id is written instead.
However using @conditional, complete object subgraphs are skipped leading to a different class enumeration such that an object cannot be decoded correctly.

**restriction 1: ** @Conditional only works if the skipped object subgraph contains preregistered classes.

Similar issues arise with shared object references: if an object is serialized twice, only a pointer to the stream position of first occurence of an object is written. However if the object subgraph skipped by @Conditional wrote the one and only instance of an object, a later backreference cannot be ressolved.

therefore

**Restriction 2:** @Conditional can be used safely only for non-shared object graphs.

I need to update documentation or drop this feature completely (as said: it has been introduced in a very early version of fst and has been used productive only for non-shared, simple flat messaging structures.

Now for the good news:

Originally, @Conditional was introduced to speed up reading in case. However for your use case it looks you just want to transform the object graph. So if you actually read all object from the stream, but don't save it to the fields, everything will work out. Its just you do not get a speed up in any way.

**Solution** (as far I understood your issue)

There is a hook to delegate serializer lookup. You then write the object using default, read the object normally but return a NULL object to avoid the result being stored. This can slow down a bit however it might solve your problem.

see https://github.com/RuedigerMoeller/fast-serialization/blob/master/src/test/ser/serializerdelegate/SerializerDelegateTest.java

Note: You'll have to clone the "master" (default) and use "mvn install" to test it (do not run tests, they take forever ..).

Pls feedback in time so I can push 2.34 release to maven.org

**Comment by Bram80** [comment]
First of all thank you for your fast response and detailed explanation!

I cloned the master branch to get your latest changes and 'mapped' your gist-example to my case but I get some exception unfortunately.

So what did I change:

global config ->
![image](https://cloud.githubusercontent.com/assets/5504516/8573686/46c37bd0-2593-11e5-8689-ce7e6821fcff.png)

My custom Serializer:
![image](https://cloud.githubusercontent.com/assets/5504516/8573699/5cdadeb8-2593-11e5-9b8e-2b214d4d809f.png)

And my custom delegate:
![image](https://cloud.githubusercontent.com/assets/5504516/8573707/687bc8e0-2593-11e5-9195-c92f0e034af3.png)

The exception that occurs when invoking in.readObject in SkipSerializer class.
![image](https://cloud.githubusercontent.com/assets/5504516/8573737/80110402-2593-11e5-8a5e-a8244e79f47d.png)

My questions:
1) Is each Serializer called for each object separately? Because in your example you do '((WrappedObject)toWrite).getWrapped())' -> please note that I'm not aware of possible references

2) I need to be able to reset only some dedicated fields of some entities. This is only required for some classes which contains an annotation with CloningMode.INSTANCE (like shown in code extract)
In your example all the instance fields are 'nullified'. Is there a way to test on a 'field by field' basis for nullifying?

**Comment by RuedigerMoeller** [comment]
3) I see the error (failed shared ref).
You actually need to register the read object (I missed that in my sample) in case it is referenced from elsewhere in the object graph
```java
@Override
    public Object instantiate(Class objectClass, FSTObjectInput in, FSTClazzInfo serializationInfo, FSTClazzInfo.FSTFieldInfo referencee, int streamPosition) throws Exception {
        in.registerObject( in.readObject() , streamPosition, serializationInfo, referencee);
        return REALLY_NULL;
    }
```

Second you don't do any check for specific classes so your skipserializer will be selected for ANY CLASS being serialized. You should check for `instanceof HibernateProxy` in the delegate, else return null (=use default).

1) normally serializers are static "function pointers" registered by class (optionally all subclasses of a class). The delegate can override serializer selection completely and globally to the configuration used.

2) Yes, I did not completely understand your use case. Its kind of hard as serializers are per object not per field. 
Actually I would need to pass the referencing field to the delegate, so you could check for annotations like:
```java
 conf.setSerializerRegistryDelegate(new FSTSerializerRegistryDelegate() {
            @Override
            public FSTObjectSerializer getSerializer(Class cl, FSTFieldInfo referencingField) {
                if (referencingField.getField().getType().getAnnotation(@NULLME) != null ) {
                    return skippingSer;
                }
                return null;
            }
        });
```

I'll check back if I could extend the SerializerRegistryDelegate interface this way .. btw expect this to have a big performance penalty ..

**Comment by RuedigerMoeller** [comment]
Nope, previous idea won't work as the serializer is cached at class level (lookup is done once per configuration). So you cannot choose based on referencing field only. Changing this would have severe performance impact for all other users.

A completely different approach:
just deserialize normally, override in FSTObjectInput.java

```java
    public Object readObject(Class... possibles) throws Exception {
            Object res = super.readObject(possibles);
            if ( res != null ) { // just manually NULL out fields marked with an annotation @MyNullMeAnnotation
                FSTClazzInfo.FSTFieldInfo[] fieldInfo = getConf().getClassInfo(res.getClass()).getFieldInfo();
                for (int i = 0; i < fieldInfo.length; i++) {
                    FSTClazzInfo.FSTFieldInfo fstFieldInfo = fieldInfo[i];
                    if ( fstFieldInfo.getField() != null && fstFieldInfo.getField().getAnnotation(MyNullMeAnnotation.class) ) {
                        fstFieldInfo.setObjectValue(res,null);
                    }
                }
            }
            return res;
    }
```

unwrapping the hibernate object can be done with the delegate, nulling out fields can be done like outlined above by simple post processing of a read object.

**Comment by RuedigerMoeller** [comment]
In order to avoid the penalty of getAnnotation for each field ref being written, king of "user field" could be added to FSTFieldInfo to speed up things, but still ..

**Comment by RuedigerMoeller** [comment]
other possibility would be to add a hook called upon each field read, passing the referencing field and object to be written. You can check for an annotation and choose to not set the field value.

**Comment by Bram80** [comment]
Thank you for your investigation and comments!

Unfortunately, I'm still having the issue of undefined reference. Also after invoking the register method.
The hook of 'readObject' in FSTObjectInput doesn't satisfy because I'm only able to inspect the fields of the root object of my object graph.

Maybe I was not clear about what I'm trying to do:

First I retrieve an object graph from DB. The point here is that it needs to be duplicated (cloned) via Serialization. Some objects need to interpreted as 'new' object in DB; therefore it's ID value (that maps with some primary key) needs to be cleared. Some other object don't need to be treated as 'new' in DB; therefore the default serialization mechanism can be used which doesn't touch it's values.

The difference between these 2 concepts (clear ID or don't clear ID) is basically made by a custom Annotation named  (like the examples in previous comments) having a value 'CloningMode.INSTANCE' (=clear the ID) or 'CloningMode.REFERENCE' (don't clear the ID).

So it seems that I need to apply the default mechanism for object with annotation value CloningMode.REFERENCE and a special mechanism (or hook) for objects with annotation value CloningMode.INSTANCE.

That's the reason why I returned the 'SkipSerializer' from the delegate in case of CloningMode.INSTANCE and 'null' in case of CloningMode.REFERENCE.

Also note that in case of CloningMode.INSTANCE I know up front which field name I have to touch.
Additionally, the Hibernate proxy thing is just to make sure that serialization begins with the 'real' object instead of the proxy. But it doesn't play a role in deciding how the object should be treated. 
An other thing: objects of INSTANCE and REFERENCE also have cyclic references with each other

But it seems that:
1) The SkipSerializer is having an undefined referencee
2) The suggested 'hook' of FSTObjectInput is not satisfying because only root object is passed.

**Comment by RuedigerMoeller** [comment]
Hm .. best thing would be to have kind of a toy structure standalone test program with a short description of what you want to achieve. I am always interested in extending applicability of fst, but probably I'd need an example to introduce appropriate hooks (however I won't sacrifice too much of performance), its just too hard to understand in detail what you want to achieve. The serializer does traverse each object so in theory it should be possible to apply any transformation required.

An interesting option would be introduction of a traversal codec (just traverses/copies objects instead of writing them). somewhat work+time required..

**Comment by Bram80** [comment]
You're correct that the serializer traverses each object.

However, if I want to hook that method I need to override FSTObjectInput::void readObjectFields
This method has package visibility.

Any chance to make it protected? It would definitely help me out!

**Comment by RuedigerMoeller** [comment]
No problem, however notice there are several variants depending on decoder (json, ..). afair there is also a branch if an object requires JDK-ser compatibility ..
I'll make everything protected as its impossible to predict people requirements in advance.

**Comment by RuedigerMoeller** [comment]
will be released with 2.35

**Comment by RuedigerMoeller** [comment]
released relax of access modifiers, hope it helps ;)
